toolchain/sem_ir/typed_insts.h

```diff
@@ -55,10 +55,8 @@ struct AdaptDecl {
 
 // The `&` address-of operator, as in `&lvalue`.
 struct AddrOf {
-  static constexpr auto Kind =
-      InstKind::AddrOf.Define<Parse::PrefixOperatorAmpId>(
-          {.ir_name = "addr_of",
-           .constant_kind = InstConstantKind::Conditional});
+  static constexpr auto Kind = InstKind::AddrOf.Define<Parse::NodeId>(
```

Comment:
I don't think I understand what you mean by "manual creation" versus "regular code" here.

The purpose of having a specific typed node associated with an instruction was as a guarantee to consumers of that instruction that the corresponding node will always be of the specified kind, so if we make changes that mean that we can't guarantee a particular node kind, we would need to change the kind to something more general, and perhaps to `NodeId`. If we have to do that on all instructions, then we should remove the typed node type entirely as it would serve no purpose.

For metaprogramming in particular: I think it's sufficiently unclear how we're going to approach metaprogramming that it's hard to speculate about what impact it might have. Maybe we'll want a new kind of `LocId` for locations produced by metaprograms, and maybe we will be able to map back to a node of the right kind in the metaprogram (depending on how we design the metaprogramming facilities).

Reply:
I mean really "created implicitly in handlers based on parse tree nodes" versus "created based on valid code". i.e., you mention the possibility of a new kind of `LocId` for metaprograms: what if you think of desugaring implicit conversions the same way? (note, I don't intend for this to block the PR: at most, maybe a note about the typical node, to make it easier to put back)

Note, maybe this is a little tangential, and I don't mean to block the PR with it, but I think the typed nodes on instructions is at an awkward point for me, neither used often enough that it's reliably there, nor ambiguous enough that I feel confident to say we can just remove it. This PR is putting more weight on the latter side though, where maybe it's not worth the work to maintain it for the nodes that _do_ have it.

Reply:
Added a comment for the two cases where we now have a less precise parse node saying what the common case is -- that seems useful information to retain, even though we have no guarantee that the parse node will actually be the one we mention.

I'm not sure that the distinction you're describing between implicitly vs explicitly created instructions is one that we should spend a lot of effort modeling in SemIR, given the decision in #3833 that SemIR should represent a desugared form of a program as if the desugared form were written directly. It might make sense to track more information about how / why we created a given instruction if it's useful for diagnostic purposes to be able to point to a different location -- and I think that might justify doing something specific for metaprograms -- but it's not clear to me that tracking which form of desugaring produced an instruction in general would pay off. Maybe we can look more into this if we find we're producing confusing diagnostics.

(And yeah, I share your concern that the typed node information might not really be pulling its weight in general.)